### PR TITLE
Gh input repo

### PR DESCRIPTION
**Pull request by sjezewski**
This PR implements #2414 (specifically the first iteration), which includes:

- the ability to add a GH repo by URL to a pipeline
- the ability to specify a name / branch on this input repository
- the ability to subscribe this repository to new commits that come into the GH repo

### Current UX:

User adds a `Github` type of input to their pipeline spec, e.g.

```
"input" : {
	"github" : {
		"URL":  "https://github.com/pachyderm/pachyderm.git",
		"Name": "my-fork",
                "Branch": "sean"
	},
}
```

And then on the [webhooks settings page for their GH repo](https://github.com/pachyderm/pachyderm/settings/hooks) they would add the URL to the githook service. This service is deployed alongside pachd, and so would be the public IP address (or load balancer) for the k8s cluster, on the right port, e.g. `1.2.3.4:30999/v1/handle/push`. 

### Implementation:

1) Pipeline Creation

When the pipeline is created, the `GithubInput` is validated (including the URL schema). What's novel here is that we create the _input_ repo at pipeline creation time, if it does not already exist.

2) Githook Server

When a new push happens on the GH repo, GH issues a POST to this githook server, which then creates a new commit on the underlying input repo. That commit is just the raw payload of the github POST. [An example can be seen here](https://github.com/pachyderm/pachyderm/blob/gh_input_repo/etc/testing/artifacts/githook-payloads/branch.json). We record this information in `commit.json`. This will enable the UI to display metadata and use the API links within that payload to semantically link back to the commit on github.

3) Worker Data Download

When the worker is downloading all the input data, it does so normally w the underlying pachyderm input repo. Once that data is loaded, it parses the payload located at `/commit.json`, and uses the `payload.Repository.CloneURL` to clone the repository locally.

Note: This PR includes a bunch of new stuff under `/vendor` because the library we use is pure go for the git client. It seems like there are a lot of golang versions of system libraries that this client depended on. I benchmarked building this branch vs building master, and it doesn't seem to have any appreciable difference on build times. (`docker make-build` takes right around 58s in either case for me)

### Limitations:

- does NOT work with private repositories
    - if a user tries to hook in a private repository, when the githook server gets the POST payload, 
    - we fail the pipeline and put a descriptive error in the `reason` field
- will clone whole repository (full depth)
    - the reason being is that there's no way w git to clone at a SHA in a single command
    - so we need to clone then checkout ... but that creates a race if the user issues another push to this branch while we're processing the first push
    - as a result, clones can take a little while
    - we have some choices / optimizations we can make here, but this is just a first pass

### Next Steps:

- explore k8s components that would let us expose a public IP for this service
- explore options (oauth app vs github app) to integrate this better for users
    - ideally we'd like the user to be able to add the github URL to their pipeline and be done (if they're authenticated via github)
    - so that on the UI, this is very simple to add via the pipeline builder
- it would be nice to prevent users from making commits to the pachyderm repo subscribed to a GH repo
    - perhaps we can create a 'machine user' for the pach cluster ... and only that one and admins have write access to this repo?
    - I do think there's a way to re-use our auth capabilities to solve this problem

### CODE REVIEW

File: etc/proto/run [hunk 1]
```diff
@@ -1,7 +1,5 @@
 #!/bin/sh
 
-set -x
```
Comment (sjezewski) [inline etc/proto/run]: This was a small unrelated change. Basically this omits the large output from `make proto` so that you can see errors. I've been bitten by this a few times, so it seemed handy to include. If we want to omit it, that's ok.

File: src/client/pps/util.go [hunk 2]
```diff
@@ -59,16 +61,59 @@ func InputCommits(input *Input) []*pfs.Commit {
 	VisitInput(input, func(input *Input) {
 		if input.Atom != nil {
 			result = append(result, &pfs.Commit{
-				Repo: &pfs.Repo{input.Atom.Repo},
+				Repo: &pfs.Repo{Name: input.Atom.Repo},
```
Comment (sjezewski) [inline src/client/pps/util.go]: This change doesn't need to be here ... let me fix these.

File: src/server/pfs/server/driver.go [hunk 3]
```diff
@@ -1023,7 +1023,13 @@ func (d *driver) subscribeCommit(ctx context.Context, repo *pfs.Repo, branch str
...
 				// We don't want to include the `from` commit itself
-				if !(seen[commit.ID] || (from != nil && from.ID == commit.ID)) {
+
+				// TODO we're check the branchName because right now WatchOne,
+				// like all collection watching commands returns prefixes which
+				// means we'll get back `master-v1` if we're looking for
+				// `master` once this is changed we should remove the
+				// comparison between branchName and branch.
+				if path.Base(branchName) == branch && (!(seen[commit.ID] || (from != nil && from.ID == commit.ID))) {
 					break
```
Comment (sjezewski) [inline src/server/pfs/server/driver.go]: This seems like a weird merge gone wrong?

AFAICT, on master it looks like the right hand side. [See here](https://github.com/pachyderm/pachyderm/blob/master/src/server/pfs/server/driver.go#L1027)

File: src/client/pps/pps.proto [hunk 4]
```diff
@@ -68,11 +68,19 @@ message CronInput {
   google.protobuf.Timestamp start = 5;
 }
 
+message GithubInput {
+  string name = 1;
+  string url = 2 [(gogoproto.customname) = "URL"];
+  string branch = 3;
+  string commit = 4;
+}
```
Comment (jdoliner) [inline src/client/pps/pps.proto]: This should be called a Git input, even if it only works for Github right now eventually we'll want it to work for all git repos. Also I was under the impression that it would already work for other git hosts. Is that true?

Comment (sjezewski) [reply src/client/pps/pps.proto]: [The library I used to handle the webhooks supports bitbucket and gitlab](https://github.com/go-playground/webhooks/tree/v3.4.0) but for us to support these we'd need to add those handlers to the githook server. Since there was no ask for those, I didn't add them. 

That said, I can change the name.

File: src/client/pps/util.go [hunk 5]
```diff
@@ -69,6 +73,60 @@ func InputCommits(input *Input) []*pfs.Commit {
 				ID:   input.Cron.Commit,
 			})
 		}
+		if input.Github != nil {
+			result = append(result, &pfs.Commit{
+				Repo: &pfs.Repo{RepoNameFromGithubInfo(input.Github.URL, input.Github.Name)},
```
Comment (jdoliner) [inline src/client/pps/util.go]: This should work like the other inputs, ie `input` should have `Repo` field on it that you use directly here and fill in, in `setPipelineDefaults`.

File: src/client/pps/util.go [hunk 6]
```diff
@@ -69,6 +73,60 @@ func InputCommits(input *Input) []*pfs.Commit {
...
+	if name != "" {
+		return name
+	}
+	// We know URL looks like:
+	// "https://github.com/sjezewski/testgithook.git",
+	tokens := strings.Split(url, "/")
+	last := tokens[len(tokens)-1]
+	tokens = strings.Split(last, ".")
+	return tokens[0]
+}
```
Comment (jdoliner) [inline src/client/pps/util.go]: Again I think it's better to use `Git` instead of `Github` for the name here. Just to leave things generic. Also it's weird to me that this takes a `name` argument just to check if it's `""`, I think the caller of this function should just do that.

File: src/client/pps/util.go [hunk 7]
```diff
@@ -69,6 +73,60 @@ func InputCommits(input *Input) []*pfs.Commit {
...
+
+// ValidateGithubCloneURL returns an error if the provided URL is invalid
+func ValidateGithubCloneURL(url string) error {
```
Comment (jdoliner) [inline src/client/pps/util.go]: Git instead of Github. Unless this only applies to github clone urls, but I don't think that's the case.

File: src/server/cmd/pachd/main.go [hunk 8]
```diff
@@ -338,6 +339,10 @@ func doFullMode(appEnvObj interface{}) error {
 		return http.ListenAndServe(fmt.Sprintf(":%v", pfs_server.HTTPPort), httpServer)
 	})
 	eg.Go(func() error {
+		err := githook.RunGitHookServer(address, etcdAddress, appEnv.PPSEtcdPrefix)
+		return err
```
Comment (jdoliner) [inline src/server/cmd/pachd/main.go]: Consolidate these into 1 line.

File: src/client/pps/util.go (hunk 6 above)
Comment (sjezewski) [reply src/client/pps/util.go]: I'll change the name.

This helper is used in a bunch of places in this PR. Yea, the caller could check, but this seemed DRYer. Also to note ... it seems like it could've taken the `input.Git` as the argument ... however this is also used on the worker side where there is the worker `input` type where this wouldn't work.

File: src/client/pps/pps.proto (hunk 4 above)
Comment (jdoliner) [reply src/client/pps/pps.proto]: Yeah, there's no need to implement those now, but particularly for protos we should always try to minimize the chances that we'll have to rename something since that causes all sorts of migration headaches.

File: src/client/pps/util.go (hunk 6 above)
Comment (jdoliner) [reply src/client/pps/util.go]: Actually after reading more I don't think you actually need this helper at all. You should just implement this behavior in `setPipelineDefaults` that way you set the `Name` field once when the pipeline comes in and from that point forward you can just use it directly.

File: src/server/pps/server/api_server.go [hunk 9]
```diff
@@ -1311,6 +1323,15 @@ func (a *apiServer) CreatePipeline(ctx context.Context, request *pps.CreatePipel
 				visitErr = err
 			}
 		}
+		if input.Github != nil {
+			if err := pps.ValidateGithubCloneURL(input.Github.URL); err != nil {
+				visitErr = err
+				return
+			}
```
Comment (jdoliner) [inline src/server/pps/server/api_server.go]: This should happen in `validateInput`.

File: src/server/pps/server/githook/githook.go [hunk 10]
```diff
@@ -0,0 +1,184 @@
...
+			}
+		}()
+	}
+}
+
+func getBranch(fullRef string) string {
+	// e.g. 'refs/heads/master'
+	tokens := strings.Split(fullRef, "/")
+	return tokens[len(tokens)-1]
+}
```
Comment (jdoliner) [inline src/server/pps/server/githook/githook.go]: This function is equivalent to `path.Base` I believe.

File: src/server/pps/server/githook/githook.go [hunk 11]
```diff
@@ -0,0 +1,184 @@
...
+					err = s.client.DeleteCommit(repoName, commit.ID)
+					return
+				}
+				err = s.client.FinishCommit(repoName, commit.ID)
+				//Final deferred function deals w non nil error
+			}()
+			err = s.client.DeleteFile(repoName, commit.ID, "commit.json")
+			if err != nil {
+				return
+			}
```
Comment (jdoliner) [inline src/server/pps/server/githook/githook.go]: Check this error inline.

File: src/server/pps/server/githook/githook.go [hunk 12]
```diff
@@ -0,0 +1,184 @@
...
+			_, err = s.client.PutFile(repoName, commit.ID, "commit.json", bytes.NewReader(raw))
+			if err != nil {
+				return
+			}
```
Comment (jdoliner) [inline src/server/pps/server/githook/githook.go]: Same here.

File: src/server/pps/server/githook/githook.go [hunk 13]
```diff
@@ -0,0 +1,184 @@
...
+		return true
+	}
+	if inputBranch == "" && payloadBranch == "master" {
+		return true
+	}
+	return false
+}
+
+func (s *gitHookServer) findMatchingPipelineInputs(payload github.PushPayload) (pipelines []*pps.PipelineInfo, inputs []*pps.GithubInput, err error) {
+	var walkInput func(*pps.Input)
```
Comment (jdoliner) [inline src/server/pps/server/githook/githook.go]: What's this function doing?

File: src/server/pps/server/githook/githook.go [hunk 14]
```diff
@@ -0,0 +1,184 @@
...
+			var inputs []*pps.Input
+			if input.Cross != nil {
+				inputs = input.Cross
+			}
+			if input.Union != nil {
+				inputs = input.Union
+			}
+			for _, input := range inputs {
+				walkInput(input)
+			}
```
Comment (jdoliner) [inline src/server/pps/server/githook/githook.go]: I don't understand what this code is doing at all.

File: src/server/pps/server/githook/githook.go [hunk 15]
```diff
@@ -0,0 +1,184 @@
...
+}
+
+func (s *gitHookServer) HandlePush(payload interface{}, header webhooks.Header) {
+	var err error
+	defer func() {
+		// handle any return error
+		if err != nil {
+			logrus.Errorf("github webhook failed to handle push with error: %v\n", err)
+		}
+	}()
```
Comment (jdoliner) [inline src/server/pps/server/githook/githook.go]: This code is badly broken, it will never log an error because `err` is shadowed below. You can see this more easily in this snippet: https://play.golang.org/p/ALqJc-_MF0 . This should be done with an inline function that returns an error instead.

Warning bells should go off in your head if you find yourself writing something like this that's not a pattern you'll find anywhere else in the code base. Chances are there's a reason.

File: src/server/pps/server/githook/githook.go [hunk 16]
```diff
@@ -0,0 +1,184 @@
...
+				loopErr = someErr
+			}
+		}
+		err = loopErr
+		return
+	}
+	triggeredRepos := make(map[string]bool)
+	for _, input := range gitInputs {
+		repoName := pps.RepoNameFromGitInfo(input.URL, input.Name)
+		if alreadyTriggered := triggeredRepos[repoName]; alreadyTriggered {
```
Comment (jdoliner) [inline src/server/pps/server/githook/githook.go]: This can just be `if triggeredRepos[repoName] {`

File: src/server/pps/server/githook/githook.go [hunk 17]
```diff
@@ -0,0 +1,184 @@
...
+			// This input is used on multiple pipelines, and we've already
+			// committed to this input repo
+			continue
+		}
+		branchName := "master"
+		if input.Branch != "" {
+			branchName = input.Branch
+		}
```
Comment (jdoliner) [inline src/server/pps/server/githook/githook.go]: You shouldn't be implementing default values here, implement defaults in one place.

File: src/server/pps/server/githook/githook.go [hunk 18]
```diff
@@ -0,0 +1,184 @@
...
+		func() {
+			var err error
+			defer func() {
+				// Handle any return error
+				if err != nil {
+					logrus.Errorf("github webhook failed to handle push with error: %v\n", err)
+				}
+				triggeredRepos[repoName] = true
+			}()
```
Comment (jdoliner) [inline src/server/pps/server/githook/githook.go]: Same problem as above, `err` is going to always be `nil` here.

File: src/server/pps/server/githook/githook.go (hunk 14 above)
Comment (sjezewski) [reply src/server/pps/server/githook/githook.go]: Ah oops, thats a leftover from where I made my own recursive walker before I found the `VisitInput` function.

File: src/server/pps/server/githook/githook.go (hunk 15 above)
Comment (sjezewski) [reply src/server/pps/server/githook/githook.go]: No, it works, because I was careful in not using inline if / error statements, so it wasn't shadowed. Here's an example of how the code is written as-is: https://play.golang.org/p/XcgwDbm39W

That said, it does feel brittle. I felt gross in particular writing the bits around line 111+

It's not often in golang, much less our codebase, that we have functions that don't return an error _and_ that we want to capture any internal errors for logging. So yes, I agree it's a novelty.

To be clear on your suggestion -- you're saying wrap the entire function body in an inline function that reports an error so we can more idiomatically return errors? I like that. Though inlining it seems a bit silly (I don't like nesting if I can help it) ... I'll probably make a non public helper function but do the same thing.

Comment (jdoliner) [reply src/server/pps/server/githook/githook.go]: My mistake.
So if you do `a, err := Err()` then that actually doesn't shadow a previous declaration of `err`... but will shadow a previous declaration of `a`. That's some pretty wild behavior.
And yeah, I was imagining you'd just wrap the whole body, but actually I think there's an even better way, just make `HandlePush` return an error, then register it with:

```go
hook.RegisterEvents(func(payload interface{}, header webhooks.Header) { 
  if err := s.HandlePush(payload, header); err != nil {
    // log err
  }
}, github.PushEvent)
```

You could also define a second function instead of inlining that, I tend to like inlining if the function isn't going to be used anywhere else but up to you.

Comment (sjezewski) [reply src/server/pps/server/githook/githook.go]: Yea it's an odd quirk for sure. And so bad to rely on. I'm glad we're cleaning this up.

I like that method of inlining, so I ran w that.

File: src/server/pps/server/api_server.go [hunk 19]
```diff
@@ -1386,6 +1398,15 @@ func (a *apiServer) CreatePipeline(ctx context.Context, request *pps.CreatePipel
 				visitErr = err
 			}
 		}
+		if input.Git != nil {
+			if err := pps.ValidateGitCloneURL(input.Git.URL); err != nil {
+				visitErr = err
+				return
+			}
```
Comment (jdoliner) [inline src/server/pps/server/api_server.go]: This should happen in `validateInput` otherwise we might create other input repos and then have to bail out when we realize this one is malformed.

File: src/server/pps/server/githook/githook.go [hunk 20]
```diff
@@ -0,0 +1,159 @@
...
+		hook,
+		c,
+		etcdClient,
+		ppsdb.Pipelines(etcdClient, etcdPrefix),
+	}
+
+	hook.RegisterEvents(
+		func(payload interface{}, header webhooks.Header) {
+			if err := s.HandlePush(payload, header); err != nil {
+				pl := payload.(github.PushPayload)
```
Comment (jdoliner) [inline src/server/pps/server/githook/githook.go]: Could we capture / check the `ok` parameter here? I'm guessing this library guarantees that the value it gives to this function will be of type `PushPayload` but given that it's not reflected in the type I think we should be defensive about it since if it's a different type this will segfault.

Comment (jdoliner) [reply src/server/pps/server/githook/githook.go]: You could also make `HandlePush` take a `github.PushPayload` so you only need to do the cast once.

**Comment by jdoliner** [comment]
Also, just occurred to me that we merged this without docs which is fine although generally something we should do. Can you make an issue / other pr to add docs for this when you have a chance?

**Comment by sjezewski** [comment]
Just to capture what we chatted about --- I planned on doing docs after the next iteration. The reason being ... the docs I'd have to write now are much longer than what I'd write after we have a first class k8s config that exposes this service on a public IP. Once we have that, I'll add docs, and they'll be more concise.

**Comment by thedrow** [comment]
It is possible to clone just one commit with Git>2.5 (which is pretty old).

See https://stackoverflow.com/a/30701724/85140

**Comment by sjezewski** [comment]
Good to know. It didn't seem like the git client we're using here supported
it, but maybe I missed a config option somewhere. As we hit bottlenecks,
this will be a good thing to keep in mind.
